presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java

```diff
@@ -47,31 +49,37 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
 
     public void updateMemoryUsage(long bytesAdded)
     {
-        SettableFuture<?> future;
-        synchronized (this) {
-            bufferedBytes.addAndGet(bytesAdded);
-
-            // if we are full, then breakout
-            if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone()) {
-                return;
+        long bufferedBytes = this.bufferedBytes.addAndGet(bytesAdded);
+        // detect the transition from above to below the full boundary
+        if (bufferedBytes <= maxBufferedBytes && (bufferedBytes - bytesAdded) > maxBufferedBytes) {
+            SettableFuture<?> future;
+            synchronized (this) {
+                //  if we have no callback waiting, return early
+                if (notFullFuture == NOT_FULL) {
```

Comment:
recheck `if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone())`

Reply:
It's necessary to *not* re-check inside of the synchronized block in order to preserve existing semantics. Re-checks can race with concurrent modifications whereas previously all modifications were serialized through the object monitor. The effect was that previously the transition edge from above-full to below full always unblocked any waiters, and not doing so here (even if we've transitioned back up above the full boundary in the mean time) is a departure from that behavior.